### ISSUE TITLE
Set up more robust site rendering scheme

### ISSUE THREAD

**Issue by andrewpbray** [andrewpbray/quarto-plotly#1]
The plan is to use this small repo to test out a new site rendering scheme for the larger Stat 20. After getting it working here, we can transition stat20/course-materials over to use it.

Working through the different components bit by bit:

- [x] 1. Set up netlify to detect changes in the repo and automatically publish from `_site/` on `main`.
- [x] 2. Set up GHA to `quarto render` site and commit new files in `_site/` to `main`.
- [x] 3. Use `renv` to be sure that local and CI calls of `quarto render` can use the same compute environment.
- [ ] 4. Shift `_site/` to a separate branch and direct netlify to monitor that one.

**Comment by andrewpbray** [comment]
@ryanlovett I was able to get 2. working (a-ok)[https://github.com/andrewpbray/quarto-plotly/actions/runs/4546394345] but then I added a sample lecture notes that is rendered both to html and to pdf. This action (didn't go through)[https://github.com/andrewpbray/quarto-plotly/actions/runs/4546577164/jobs/8015379109]; there was an issue with the tex install. I tried to find the description of the image to see if tex could be updated there, but I couldn't find it. Any ideas?

**Comment by ryanlovett** [comment]
@andrewpbray I think a strategy to solve that action is to install a more fully functional TeX environment in the docker image. This is what we did for stat20 and other datahub environments that relied on TeX. It was easier to preinstall than to have the resources be pulled on demand. I'll update the docker image then you should be able to rerun the workflow. The current action references `ucbscf/stat20-course-materials:latest` which, in production, is not a good idea since it doesn't fix the version. In this context however, pulling :latest will fetch the latest image which is handy.

**Comment by ryanlovett** [comment]
I updated the image and that action now passes.

**Comment by andrewpbray** [comment]
@ryanlovett Ah, that makes sense to me. Where can I find `ucbscf/stat20-course-materials:latest`? That's the file where can specify which r packages are being used, right?

**Comment by andrewpbray** [comment]
Also, what are your thoughts on https://rstudio.github.io/renv/articles/docker.html?

This is regarding step 3 above.

**Comment by ryanlovett** [comment]
The Dockerfile and related files that are used to build `ucbscf/stat20-course-materials:latest` are in https://github.com/ryanlovett/quarto-plotly. They can be moved to your repo, in an `image` subdirectory or some such.

That base image can be converted to use `renv`, instead of using custom .r files to install packages.

**Comment by andrewpbray** [comment]
Sorry I'm obtuse here, but how do these lines in the workflow file point it to `/image/DOCKERFILE`?

```
    container:
      image: ucbscf/stat20-course-materials:latest
```
That's to say, what's the the use of the `ucbscf/stat20-course-materials:latest` string?  I don't see that references anywhere in `image`. Is it just a name that is now attached to the image? What's the use of naming it like that?

**Comment by ryanlovett** [comment]
Ah, no worries. To build the docker image I changed to the directory containing the Dockerfile and ran `docker build -t ucbscf/stat20-course-materials:latest .`. When it successfully completed I ran `docker push ucbscf/stat20-course-materials:latest` to push it up to docker hub. This last step uses credentials associated with the ucbscf docker hub org, though we can eventually make a separate org for this, e.g. "stat20".

Github Actions (or anyone) can then pull that image down.

Ideally, the docker build/push steps would be part of the CI workflow so that whenever a change is made to the Dockerfile or related files, it automatically gets built, tagged with the SHA of the git commit instead of "latest", and pushed.

**Comment by andrewpbray** [comment]
Ah, that explains it!  I didn't know there was such a thing as docker hub. I was looking for `github.com/ucbscf/stat20-course-materials` and not finding anything. And that makes sense about how building and pushing that image could be made part of CI.

**Comment by andrewpbray** [comment]
So to accomplish #3, for now I could just add packages to `https://github.com/andrewpbray/quarto-plotly/blob/master/image/bray-base.r`, correct? Then down the road we can convert the docker image to use renv if needed?

**Comment by ryanlovett** [comment]
Exactly!

**Comment by andrewpbray** [comment]
I just pushed a commit that adds `stat20/stat20data` to `bray-base.r`, and it didn't work. I realize why now: the action is configured to look to docker hub for that file. Is there a way to get it to just look into the `image` directory in this repo?

**Comment by ryanlovett** [comment]
Once the Dockerfile and/or related files have changed, we have to rebuild the image and push it to docker hub. Docker hub won't build images for us.

I just pulled your latest changes, ran `docker build ...` as specified above, then `docker push`. 

I'll try to prepare automated docker image builds from GA this week. I can also show you how to do it manually from your laptop. It is worthwhile to learn because iterating locally is often necessary to resolve bugs, rather than waiting for GA to run and fail.

I just re-ran CI and it failed due to the image not having `patchwork`. I've added it and will rerun.

**Comment by ryanlovett** [comment]
Another change, and it [succeeded](https://github.com/andrewpbray/quarto-plotly/actions/runs/4593250481/jobs/8110998809).

**Comment by andrewpbray** [comment]
I went ahead and applied for a free OSS org/account on Docker Hub so that we can have a `stat20` org. I'll let you know when I hear back.

Do you have any clever ideas for how to do 4? The main challenge I see has to do with .gitignore files. The reason for shiftng the `_site/` to a separate `netlify` or `website` branch would be to keep `main` clean: no `_freeze`, no `_site`; one source. So I'd like to add both `_freeze` and `_site` to the .gitignore. But if I do that, then how do I commit those files through CI on the `website` branch?

I came across this solution, but was wondering if you had any other ideas that might be simpler, since really this is just trying to implement `gh-pages` but with netlify.

https://gist.github.com/wizioo/c89847c7894ede628071

**Comment by ryanlovett** [comment]
One can checkout branches into separate working directories and send output to the desired branch. There are probably cleverer ways to do this without multiple working directories.

Can you please add `ryanlovett` to your `stat20` docker hub org? I can start pushing images to it.

**Comment by andrewpbray** [comment]
Yep, I'll add you as soon as I figure out how to get access to it (I've sent them a follow-up email).

About the checking out into multiple dirs: how does would we actually commit the output to the desired branch if there's a .gitignore file? Could you demo it here in this repo by copying an modifying the existing GHA?
